third_party/terraform/utils/provider_test.go.erb

```diff
@@ -96,6 +106,88 @@ func init() {
 	<% end -%>
 }
 
+var lock = &sync.Mutex{}
+
+// Returns a cached config if VCR testing is enabled. This enables us to use a single HTTP transport
+// for a given test, allowing for recording of HTTP interactions.
+func getCachedConfig(d *schema.ResourceData, configureFunc func(d *schema.ResourceData) (interface{}, error), testName string) (interface{}, error) {
+	lock.Lock()
+	defer lock.Unlock()
+	if _, ok := configs[testName]; !ok {
+		c, err := configureFunc(d)
+		if err != nil {
+			return nil, err
+		}
+		config := c.(*Config)
+		vcrMode := recorder.ModeDisabled
+		switch vcrEnv := os.Getenv("VCR_MODE"); vcrEnv {
+		case "RECORDING":
+			vcrMode = recorder.ModeRecording
+		case "REPLAYING":
+			vcrMode = recorder.ModeReplaying
+		default:
+			log.Printf("[DEBUG] No valid environment var set for VCR_MODE, skipping VCR. VCR_MODE: %s", vcrEnv)
+			return c, nil
+		}
+
+		envPath := os.Getenv("VCR_PATH")
+		if envPath == "" {
+			log.Printf("[DEBUG] No environment var set for VCR_PATH, skipping VCR for test: %s", testName)
+			return c, nil
+		}
+		path := filepath.Join(envPath, testName)
+
+		rec, err := recorder.NewAsMode(path, vcrMode, config.client.Transport)
+		if err != nil {
+			return nil, err
+		}
+		rec.SetMatcher(func(r *http.Request, i cassette.Request) bool {
+			if r.Body == nil {
+				return cassette.DefaultMatcher(r, i)
+			}
+			var b bytes.Buffer
+			if _, err := b.ReadFrom(r.Body); err != nil {
+				return false
+			}
+			r.Body = ioutil.NopCloser(&b)
+			return cassette.DefaultMatcher(r, i) && (b.String() == "" || b.String() == i.Body)
+		})
+		config.client.Transport = rec
+		configs[testName] = c
+		return c, err
+	}
+	return configs[testName], nil
+}
+
+// We need to explicitly close the VCR recorder to save the cassette
+func closeRecorder(testName string) {
+	if config, ok := configs[testName]; ok {
+		// We did not cache the config if it does not use VCR
+		config.(*Config).client.Transport.(*recorder.Recorder).Stop()
+		// Clean up test config
+		delete(configs, testName)
+	}
+}
+
+func getTestAccProviders(testName string) map[string]terraform.ResourceProvider {
+	prov := Provider().(*schema.Provider)
+	provRand := random.Provider().(*schema.Provider)
+	if v := os.Getenv("VCR_ENABLED"); v != "" {
```

Comment:
It looks like the other function skipped VCR if either of VCR_MODE or VCR_PATH were unset. Does it make sense to use those here too instead of adding a third?